src/gp_creds.c

```diff
@@ -667,6 +667,41 @@ uint32_t gp_add_krb5_creds(uint32_t *min,
             if (ret_maj) {
                 goto done;
             }
+
+            if (req_name != GSS_C_NO_NAME) {
+                int equal = 0;
+
+                ret_maj = gss_inquire_cred(&ret_min, impersonator_cred,
+                                           &target_name, NULL, NULL, NULL);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                ret_maj = gss_compare_name(&ret_min, target_name,
+                                           req_name, &equal);
+                if (ret_maj) {
+                    goto done;
+                }
+
+                /* if impersonator credential retrieval yielded the requested client
+                 * name, we do not need to impersonate. Also, in AD an attempt to
+                 * impersonate oneself yields an error "KDC has no support for padata
+                 * type" */
+                if (equal) {
+                     ret_maj = gss_acquire_cred_from(&ret_min, req_name, GSS_C_INDEFINITE,
+                                                     &desired_mechs, cred_usage,
+                                                     &cred_store, &user_cred,
+                                                     actual_mechs, NULL);
+                     if (!ret_maj) {
```

Comment:
If we end up maintaining this additional Acquire cred call though this need to be changed to if (ret_maj != GSS_COMPLETE)

Reply:
Do you mean ret_maj == GSS_S_COMPLETE (aka 0)? It's on this condition where we bail out early but successfully. I was afraid that this might be unintuitive to read because it's opposed to all the other error check logic. But I wanted to get the new logic in with as little refactoring as possible.

Also I noticed that we actually jump over initiator_time_req retrieval further down (https://github.com/gssapi/gssproxy/blob/main/src/gp_creds.c#L728). As does the other early bailout (https://github.com/gssapi/gssproxy/blob/main/src/gp_creds.c#L699). Is this a concern? Perhaps the whole logic flow needs a rethink?

Reply:
Yes we should use GSS_ERROR() or explicit comparison, do not worry about refactoring other users though.

Reply:
I've fixed the error check to read more easily using GSS_S_COMPLETE.